File: solutions/vbnet/bob/1/Bob.vb
Public Class Bob

  private function is_all_caps(statement as string) as integer
        dim letter as string
        dim code as integer
        dim contains_caps as integer
        dim contains_noncaps as integer
        for idx=0 to len(statement)-1
            letter=statement(idx)
            code=asc(letter)
            contains_caps=contains_caps or (code>=asc("A") and code<=asc("Z"))
            contains_noncaps=contains_noncaps or (code>=asc("a") and code<=asc("z"))
        next idx
        return contains_caps and (not contains_noncaps)
  end function

    private Function strip(statement As String) As String
        dim stripped_statement as string
        dim letter as string
        stripped_statement=""
        for idx=0 to len(statement)-1
            letter=statement(idx)
            if letter<>" " then
                stripped_statement=stripped_statement+letter
            end if
        next idx
        return stripped_statement
    End Function

    Public Function Hey(statement As String) As String
        dim stripped_statement as string
        dim yelled as integer
        stripped_statement=""
        if statement<>"" then
            stripped_statement=strip(statement)
        end if
        if stripped_statement="" then
            return "Fine. Be that way!"
        end if
        yelled=is_all_caps(stripped_statement)
        if stripped_statement(len(stripped_statement)-1)="?" then
            if yelled then
                return "Calm down, I know what I'm doing!"
            end if
            return "Sure."
        end if
        if yelled then
            return "Whoa, chill out!"
        end if
        return "Whatever."
    End Function
End Class

File: solutions/cpp/bob/4/bob.cpp
#include "bob.h"

namespace bob {

bool is_whitespace(char letter) {
    const int num_of_whitespace_characters = whitespace_characters.size();
    int idx;
    for (idx = 0; idx < num_of_whitespace_characters; idx++) {
        if (letter == whitespace_characters[idx]) {
            break;
        }
    }
    return idx < num_of_whitespace_characters;
}

bool is_lowercase(char letter) {
    return letter >= 'a' && letter <= 'z';
}

bool is_uppercase(char letter) {
    return letter >= 'A' && letter <= 'Z';
}

std::string hey(const std::string& greeting) {
    auto greeting_without_whitespace = remove_whitespace(greeting);
    const auto len = greeting_without_whitespace.length();
    
    if (len == 0) {
        return "Fine. Be that way!";
    }
    
    auto yelled = is_all_capitals(greeting_without_whitespace);
    
    if (greeting_without_whitespace[len - 1] == '?') {
        if (yelled) {
            return "Calm down, I know what I'm doing!";
        }
        return "Sure.";
    }
    if (yelled) {
        return "Whoa, chill out!";
    }
    return "Whatever.";
}

std::string remove_whitespace(const std::string& sentence) {
    const int len = sentence.length();
    int idx;
    for (idx = 0; idx < len; idx++) {
        if (!is_whitespace(sentence[idx])) {
            break;
        }
    }
    const auto start_pos = idx;
    for (idx = len - 1; idx >= 0; idx--) {
        if (!is_whitespace(sentence[idx])) {
            break;
        }
    }
    const auto end_pos = idx + 1;
    return sentence.substr(start_pos, end_pos - start_pos);
}

bool is_all_capitals(const std::string& sentence){
    auto contains_capitals = false;
    auto contains_low = false;
    for(const auto& letter : sentence){
        contains_capitals = contains_capitals || is_uppercase(letter);
        contains_low = contains_low || is_lowercase(letter);
    }
    return contains_capitals && !contains_low;
}

}  // namespace bob
